File: ADT/GameEngine/Domain/Question.hs
-- |
-- Question definition domain: declarative model for bootstrap + runtime rules.
module GameEngine.Domain.Question where

import Data.Text (Text)
import GameEngine.Domain.Behavior
import GameEngine.Domain.Common
import GameEngine.Domain.Entity
import GameEngine.Domain.Space

data QuestionMeta = QuestionMeta
  { questionMetaId :: QuestionId,
    questionMetaTitle :: Text,
    questionMetaDescription :: Text
  }
  deriving (Eq, Show)

data EntityDefinition = EntityDefinition
  { definitionEntityConfig :: ItemDataConfig,
    definitionInitialSpace :: Maybe SpaceId,
    definitionInitialPosition :: Maybe GridPosition
  }
  deriving (Eq, Show)

data Condition
  = CondAnd [Condition]
  | CondOr [Condition]
  | CondNot Condition
  | CondFlag Text Bool
  | CondEq Text Value
  | CondIn Text [Value]
  deriving (Eq, Show)

data PhaseRule
  = PhaseSet Condition Phase
  | PhaseRetain Condition
  deriving (Eq, Show)

data InventoryRule
  = InventoryShowGroup Condition Text
  | InventoryHideGroup Condition Text
  deriving (Eq, Show)

data SpaceRule
  = SpaceShow Condition SpaceId
  | SpaceHide Condition SpaceId
  deriving (Eq, Show)

data QuestionDefinition ctx = QuestionDefinition
  { definitionMeta :: QuestionMeta,
    definitionInitialPhase :: Phase,
    definitionSpaces :: [SpaceDefinition],
    definitionEntities :: [EntityDefinition],
    definitionPhaseRules :: [PhaseRule],
    definitionInventoryRules :: [InventoryRule],
    definitionSpaceRules :: [SpaceRule],
    definitionBehaviors :: Maybe (BehaviorDefinition ctx)
  }

File: src/components/game/engine/types/question.ts
import type {
	ConditionContext as __ConditionContext,
	PhaseResolution as __PhaseResolution,
} from "../../internal/domain/question";
import type {
	PhaseRule as __PhaseRule,
	QuestionDefinition as __QuestionDefinition,
} from "../../internal/runtime/definition/types";

type _ConditionContext<ConditionKey extends string = string> =
	__ConditionContext<ConditionKey>;
type _PhaseResolution = __PhaseResolution;
type _PhaseRule<ConditionKey extends string = string> =
	__PhaseRule<ConditionKey>;
type _QuestionDefinition<
	ConditionKey extends string = string,
	TContext = Record<string, never>,
> = __QuestionDefinition<ConditionKey, TContext>;

export type ConditionContext<ConditionKey extends string = string> =
	_ConditionContext<ConditionKey>;
export type PhaseResolution = _PhaseResolution;
export type PhaseRule<ConditionKey extends string = string> =
	_PhaseRule<ConditionKey>;
export type QuestionDefinition<
	ConditionKey extends string = string,
	TContext = Record<string, never>,
> = _QuestionDefinition<ConditionKey, TContext>;
